utils/utils.go

```diff
@@ -101,3 +101,27 @@ func WorkspaceExists(ctx context.Context, client *kong.Client) (bool, error) {
 	}
 	return client.Workspaces.Exists(ctx, &workspace)
 }
+
+func AsKongConsumer(c kong.Consumer) *kong.Consumer {
+	consumer := &kong.Consumer{ID: kong.String(*c.ID)}
+	if c.Username != nil {
+		consumer.Username = kong.String(*c.Username)
+	}
+	return consumer
+}
+
+func AsKongService(s kong.Service) *kong.Service {
+	service := &kong.Service{ID: kong.String(*s.ID)}
+	if s.Name != nil {
+		service.Name = kong.String(*s.Name)
+	}
+	return service
+}
+
+func AsKongRoute(r kong.Route) *kong.Route {
+	route := &kong.Route{ID: kong.String(*r.ID)}
+	if r.Name != nil {
+		route.Name = kong.String(*r.Name)
+	}
+	return route
+}
```

Comment:
Yeah I totally agree...what about `GetStrippedRoute` or `GetRouteWithIDAndName`?

Reply:
`GetIDNameOnlyServiceCopy` ?

The doc comments are the more important part. Something like

> // GetRouteWithIDAndName takes a kong.Route and returns a reference to a new kong.Route. The new kong.Route has the ID and Name fields from the input route, but none of the other input route's fields.

And above all such functions:

> // These functions return stripped copies (ID and Name only) of Kong resource objects. We use these elsewhere [in place] because we cannot use the original objects, as objects with all fields present would [result in some unwanted outcome].

with the `[placeholder]` bits filled in. It's not immediately clear to me why we wouldn't use the original objects (they do already have their Name and ID, so it seems like we'd just be able to use them as-is), so while I assume there is some reason, I can't intuit it, and suspect that future readers will have the same confusion.

Hijacking the end of this for a pre-release changelog entry reminder also.

Reply:
did a bit more of investigation here and it seems that we actually only need to zero-out the timestamps from the entities, instead of creating new ones only with IDs and Names https://github.com/Kong/deck/pull/662/commits/1364c151b686473fff3f74182c28979400247e83

(it's better to fully visualise it from [here](https://github.com/Kong/deck/pull/662/files#diff-d0401391a7f45651a98e489a540d95eb1c3cda5d8e077414f234f15c06a8bc3eR204))

WDYT?

Reply:
Ah, now it makes sense. That explanation was basically exactly what I was looking for in the comment :smile: I wasn't making the connection that these were references. With that in mind, the original name/ID only approach makes more sense than timestamp stripping:

```
// These GetFooReference functions return stripped copies (ID and Name only) of Kong resource structs. We use these
// within KongRawState structs to indicate entity relationships. While state files indicate relationships by nesting (A 
// collection of services is [{name: "foo", id: "1234", connect_timeout: 600000, routes: [{name: "fooRoute"}]}]), 
// KongRawState is flattened, with all entities listed independently at the top level. To preserve the relationships, these 
// flattened entities include references (the route from earlier becomes {name: "fooRoute", service: {name: "foo", id: 
// "1234"}}).

// GetRouteReference returns a name+ID only copy of the input route, for use in references from other objects
GetRouteReference() {
... return the ID/name only copy
}
```

Reply:
Totally agree, that wasn't clear to me either :)

https://github.com/Kong/deck/pull/662/commits/2b4c09d6c5f6326a827278e832a3a4e0d581bd4b